src/renderer_vk.cpp

```diff
@@ -867,6 +871,264 @@ VK_IMPORT_DEVICE
+				}
+			}
+
+			m_needToRefreshSwapchain = false;
+
+			return result;
+		}
+
+		void releaseSwapchain()
+		{
+			VK_CHECK(vkDeviceWaitIdle(m_device) );
+			vkFreeMemory(m_device, m_backBufferDepthStencilMemory, m_allocatorCb);
+			vkDestroy(m_backBufferDepthStencilImageView);
+			vkDestroy(m_backBufferDepthStencilImage);
+			for (uint32_t ii = 0; ii < BX_COUNTOF(m_backBufferColorImageView); ++ii)
+			{
+				vkDestroy(m_backBufferColorImageView[ii]);
+			}
+			vkDestroy(m_swapchain);
+		}
+
+		VkResult createSwapchainFramebuffer()
+		{
+			VkResult result = VK_SUCCESS;
+			for (uint32_t ii = 0; ii < m_numSwapchainImages; ++ii)
+			{
+				::VkImageView attachments[] =
+				{
+					m_backBufferColorImageView[ii],
+					m_backBufferDepthStencilImageView,
+				};
+
+				VkFramebufferCreateInfo fci;
+				fci.sType = VK_STRUCTURE_TYPE_FRAMEBUFFER_CREATE_INFO;
+				fci.pNext = NULL;
+				fci.flags = 0;
+				fci.renderPass = m_renderPass;
+				fci.attachmentCount = BX_COUNTOF(attachments);
+				fci.pAttachments = attachments;
+				fci.width = m_sci.imageExtent.width;
+				fci.height = m_sci.imageExtent.height;
+				fci.layers = 1;
+
+				result = vkCreateFramebuffer(m_device, &fci, m_allocatorCb, &m_backBufferColor[ii]);
+
+				if (VK_SUCCESS != result)
+				{
+					return result;
+				}
+			}
+			return result;
+		}
+
+		void releaseSwapchainFramebuffer()
+		{
+			for (uint32_t ii = 0; ii < BX_COUNTOF(m_backBufferColorImageView); ++ii)
+			{
+				vkDestroy(m_backBufferColor[ii]);
+			}
+		}
+
+		void initSwapchainImageLayout()
+		{
+			VkCommandBuffer commandBuffer = beginNewCommand();
+			VkRenderPassBeginInfo rpbi;
+			rpbi.sType = VK_STRUCTURE_TYPE_RENDER_PASS_BEGIN_INFO;
+			rpbi.pNext = NULL;
+			rpbi.renderPass = m_renderPass;
+			rpbi.renderArea.offset.x = 0;
+			rpbi.renderArea.offset.y = 0;
+			rpbi.renderArea.extent = m_sci.imageExtent;
+			rpbi.clearValueCount = 0;
+			rpbi.pClearValues = NULL;
+
+			setImageMemoryBarrier(
+				  commandBuffer
+				, m_backBufferDepthStencilImage
+				, VK_IMAGE_ASPECT_DEPTH_BIT | VK_IMAGE_ASPECT_STENCIL_BIT
+				, VK_IMAGE_LAYOUT_UNDEFINED
+				, VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL
+				, 1
+				, 1
+				);
+
+			for (uint32_t ii = 0; ii < m_numSwapchainImages; ++ii)
+			{
+				setImageMemoryBarrier(
+					  commandBuffer
+					, m_backBufferColorImage[ii]
+					, VK_IMAGE_ASPECT_COLOR_BIT
+					, VK_IMAGE_LAYOUT_UNDEFINED
+					, VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL
+					, 1
+					, 1
+					);
+
+				rpbi.framebuffer = m_backBufferColor[ii];
+				vkCmdBeginRenderPass(commandBuffer, &rpbi, VK_SUBPASS_CONTENTS_INLINE);
+				vkCmdEndRenderPass(commandBuffer);
+
+				setImageMemoryBarrier(
+					  commandBuffer
+					, m_backBufferColorImage[ii]
+					, VK_IMAGE_ASPECT_COLOR_BIT
+					, VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL
+					, VK_IMAGE_LAYOUT_PRESENT_SRC_KHR
+					, 1
+					, 1
+					);
+			}
+
+			m_backBufferColorIdx = 0;
+			submitCommandAndWait(commandBuffer);
```

Comment:
If I'm reading this right, you're doing operations on swap-chain images before they've been acquired here? From the spec:

```Use of a presentable image must occur only after the image is returned by vkAcquireNextImageKHR, and before it is presented by vkQueuePresentKHR. This includes transitioning the image layout and rendering commands.```

So this must be deferred to after the acquire.